### PR TITLE
bats: improve package; move installCheck to passthru

### PR DESCRIPTION
**Pull request by abathur**
###### Description of changes

- improve package w/ new resholve features
- move installCheck into passthru.tests

###### Things done

Result of `nixpkgs-review` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>bash-preexec</li>
    <li>bats</li>
  </ul>
</details>

Result of `nixpkgs-review` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>bash-preexec</li>
    <li>bats</li>
  </ul>
</details>

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [x] x86_64-darwin
  - [ ] aarch64-darwin
- [x] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/interpreters/bats/default.nix [hunk 1]
```diff
@@ -31,32 +37,99 @@ resholve.mkDerivation rec {
...
-    sed -i '/test works even if PATH is reset/a skip' test/bats.bats
+  passthru.tests.upstream = bats.unresholved.overrideAttrs (old: {
+    name = "${bats.name}-tests";
+    installCheckInputs = [
+      ncurses
+      parallel # skips some tests if it can't detect
+      flock # skips some tests if it can't detect
+      ps
+    ] ++ lib.optionals stdenv.isDarwin [ lsof ];
+    inherit doInstallCheck;
```
Comment (SuperSandro2000) [inline pkgs/development/interpreters/bats/default.nix:114]: All of this should not be in passtruh and only doCheck rule be overwritten.

Comment (abathur) [reply pkgs/development/interpreters/bats/default.nix:114]: This is one of the main points of the PR, so I'm inclined to back on this without sources/reasoning. 

While this approach is rare, I don't think it's unprecedented within nixpkgs. It's also roughly aligned with the WIP recommendations in https://github.com/NixOS/rfcs/pull/119/ (specifically https://github.com/NixOS/rfcs/pull/119/files#diff-d207098de17afa3aaea2ef0e420b2b1c674b257267828c11e40928351df3e368R46-R54).

Comment (SuperSandro2000) [reply pkgs/development/interpreters/bats/default.nix:114]: I mean like https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/python-modules/sortedcontainers/default.nix#L27-L29

Comment (abathur) [reply pkgs/development/interpreters/bats/default.nix:114]: That makes sense, though it doesn't work in this case. 

I may have mis-designed resholve's Nix API, but it [produces an inner derivation](https://github.com/NixOS/nixpkgs/blob/d59dd43e49f24b58fe8d5ded38cbdf00c3da4dc2/pkgs/development/misc/resholve/resholve-utils.nix#L152-L171) straightforwardly, and then [uses it as the source for an outer derivation that is passed through resholve](https://github.com/NixOS/nixpkgs/blob/d59dd43e49f24b58fe8d5ded38cbdf00c3da4dc2/pkgs/development/misc/resholve/resholve-utils.nix#L173-L202).

I'm not sure if that is at all persuasive, but resholve's Nix API and this test can be incrementally improved in subsequent PRs.

Comment (abathur) [reply pkgs/development/interpreters/bats/default.nix:114]: @SuperSandro2000 ping? thoughts? I can also bug someone else if you're short on time :)

**Pull request merged**